proposals/p1885.md

```diff
@@ -0,0 +1,458 @@
+
+```rust
+// Carbon definition
+interface Indexable() {
+    let ElementType:! Type
+    fn GetSize[me: Self]() -> u32;
+    fn GetAt[me: Self](pos: u32) -> ElementType;
+}
+
+// Usage
+class MyArray {
+    ...
+    impl as Indexable where .ElementType = i32 {
+        fn GetSize[me: Self]() -> u32 { ... };
+        fn GetAt[me: Self](pos: u32) -> ElementType { ... };
+    }
+}
+
+var array: MyArray = {...};
+for (var element: auto in array) {
+    ...
+}
+```
+
+#### Implicit `[]` support
+
+A related added value would be the support for index-based access and assignment
+with the `[]` operator, as a 1D array, for `Indexable` types. In addition to
+basic single-element operations, this opens the door for future more advance
+uses, as proposed in
+[PR-1928](https://github.com/carbon-language/carbon-lang/pull/1928).
+
+The type-checker would however not be able to validate `Indexable`-based array,
+in comparison to fixed arrays:
+
+```rust
+fn Foo[T:! Indexable](array: T) {
+    array[1] = array[2]; // May or may not exist
+}
+```
+
+#### Mutable access
+
+TODO
+
+### Consumable
+
+An interface representing single-use sequential access type of container. Items
+ca be read (consumed) once, and are not expected to be accessible anymore after
+usage.
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   For each iteration, call `Consume() -> Optional(ElementType)`
+-   If the `Optional` is null, break the loop. Otherwise return the value held
+    by the Optional.
+
+#### Interface and usage
+
+```rust
+// Carbon definition
+interface Consumable() {
+    let ElementType:! Type
+    fn Consume[me: Self]() -> Optional(ElementType);
+}
+
+// Usage
+class MyContainer {
+    ...
+    impl as Consumable where .ElementType = i32 {
+        fn Consume[me: Self]() -> Optional(ElementType) {
+            ...
+        }
+    }
+}
+
+var container: MyContainer = {...};
+for (var a: auto in container) {}
+// `container` is now empty
+```
+
+#### TODO
+
+-   Size method?
+
+### Iterable
+
+The `Iterable` interface represents a generic sequential-access type of
+container. It relies on two new interfaces:
+
+-   `Iteratable`: Interface for a type that can return an iterator and be
+    iterated over
+-   `Iterator`: Interface for an iterator-like object
+
+**Note:** This interface makes use of `Optional` type, and assumes its future
+definition. The usage highlighted here is hypothetical.
+
+#### Range-based `for` behavior
+
+-   Get an iterator with `GetIterator()`
+-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
+    -   If the `Optional` is null, break the loop. Otherwise return the value
+        held by the Optional.
+
+#### Interface and usage
+
+```rust
+// Basic immutable iterator
+interface Iterator() {
+    let ElementType:! Type;
+    fn Next[me: Self]() -> Optional(ElementType);
```

Comment:
It is definitely going to boil down to preference from an ergonomics and performance standpoint. If we put an empasis on performance, this might just not be an option. Offering a `SafeIterator` might do it, with the annoying multiplication of interfaces as a side effect (`SafeMutableIterator`?)

Reply:
Removed `Optional` and added to "Alternative considered"